File: ocaml/day07.ml
open CCFun

let instructions = Intcode.read_instructions "inputs/07.txt"

let all_permutations =
  (* https://codereview.stackexchange.com/questions/125173/generating-permutations-in-ocaml *)
  let rec aux result other = function
    | [] -> [result]
    | hd :: tl ->
      let r = aux (hd :: result) [] (other @ tl) in
      if tl <> [] then
        r @ aux result (hd :: other) tl
      else r
  in
  aux [] []


let create_computers =
  List.map
    (fun phase ->
      instructions
      |> Intcode.initialize_computer
      |> Intcode.receive phase)

let some_halted =
  List.exists (fun comp -> Intcode.get_state comp = Intcode.Halted)

let rec get_output (score, computers) =
  if some_halted computers then score
  else
    computers
    |> CCList.fold_map
      (fun last_output comp ->
           comp
           |> Intcode.receive last_output
           |> Intcode.run_until_halt
           |> fun comp -> ((Intcode.get_next_output comp), comp) )
      score
    |> get_output


let solve =
  all_permutations
  %> List.fold_left
    (fun acc perm ->
       let computers = create_computers perm in
       (0, computers) |> get_output |> max acc)
    0
  %> Printf.printf "%d\n"


let () =
  CCList.(0 -- 4) |> solve;
  CCList.(5 -- 9) |> solve

File: python/day07.py
from intcode import *
from itertools import permutations


def create_computer(instr, i):
    comp = Computer(instr[:], len(instr))
    comp.receive(i)
    return comp

def solve(instr, ranger):
    res = 0
    for perm in permutations(ranger):
        computers = [create_computer(instr, perm[i]) for i in range(5)]
        output = 0
        while all(c.state != State.Halted for c in computers):
            for c in computers:
                output = c.receive_run_output(output)
        res = max(res, output)
    return res


instructions = read_intcode_input(7)

print(solve(instructions, range(5)))
print(solve(instructions, range(5, 10)))
